cmd/kubeadm/app/phases/certs/certs.go

```diff
@@ -123,7 +123,9 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration, pkiDir string) error {
 		config := certutil.Config{
 			CommonName: "kube-apiserver",
 			AltNames:   altNames,
-			Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+			// This makes the apiserver allowed to talk to the kubelets in the cluster
+			Organization: []string{"system:masters"},
```

Comment:
@liggitt Do you think something like this will work then?
```
--- a/cmd/kubeadm/app/constants/constants.go
+++ b/cmd/kubeadm/app/constants/constants.go
@@ -24,4 +24,8 @@ const (
        APIServerCertAndKeyBaseName = "apiserver"
        APIServerCertName           = "apiserver.crt"
        APIServerKeyName            = "apiserver.key"
+
+       APIServerKubeletClientCertAndKeyBaseName = "apiserver-kubelet-client"
+       APIServerKubeletClientCertName           = "apiserver-kubelet-client.crt"
+       APIServerKubeletClientKeyName            = "apiserver-kubelet-client.key"
 )
diff --git a/cmd/kubeadm/app/master/manifests.go b/cmd/kubeadm/app/master/manifests.go
index bf44fad..b234b70 100644
--- a/cmd/kubeadm/app/master/manifests.go
+++ b/cmd/kubeadm/app/master/manifests.go
@@ -322,8 +322,8 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, selfHosted bool) [
                "--client-ca-file="+getCertFilePath(kubeadmconstants.CACertName),
                "--tls-cert-file="+getCertFilePath(kubeadmconstants.APIServerCertName),
                "--tls-private-key-file="+getCertFilePath(kubeadmconstants.APIServerKeyName),
-               "--kubelet-client-certificate="+getCertFilePath(kubeadmconstants.APIServerCertName),
-               "--kubelet-client-key="+getCertFilePath(kubeadmconstants.APIServerKeyName),
+               "--kubelet-client-certificate="+getCertFilePath(kubeadmconstants.APIServerKubeletClientCertName),
+               "--kubelet-client-key="+getCertFilePath(kubeadmconstants.APIServerKubeletClientKeyName),
                "--token-auth-file="+kubeadmapi.GlobalEnvParams.HostPKIPath+"/tokens.csv",
                fmt.Sprintf("--secure-port=%d", cfg.API.Port),
                "--allow-privileged",
diff --git a/cmd/kubeadm/app/phases/certs/certs.go b/cmd/kubeadm/app/phases/certs/certs.go
index 78d5533..4ac2a13 100644
--- a/cmd/kubeadm/app/phases/certs/certs.go
+++ b/cmd/kubeadm/app/phases/certs/certs.go
@@ -123,8 +123,6 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration, pkiDir string) error {
                config := certutil.Config{
                        CommonName: "kube-apiserver",
                        AltNames:   altNames,
-                       // This makes the apiserver allowed to talk to the kubelets in the cluster
-                       Organization: []string{"system:masters"},
                        Usages:       []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
                }
                apiCert, apiKey, err := pkiutil.NewCertAndKey(caCert, caKey, config)
@@ -138,6 +136,35 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration, pkiDir string) error {
                fmt.Println("[certificates] Generated API server certificate and key.")
        }
 
+       // If at least one of them exists, we should try to load them
+       // In the case that only one exists, there will most likely be an error anyway
+       if pkiutil.CertOrKeyExist(pkiDir, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName) {
+               // Try to load ca.crt and ca.key from the PKI directory
+               apiCert, apiKey, err := pkiutil.TryLoadCertAndKeyFromDisk(pkiDir, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName)
+               if err != nil || apiCert == nil || apiKey == nil {
+                       return fmt.Errorf("certificate and/or key existed but they could not be loaded properly")
+               }
+
+               fmt.Println("[certificates] Using the existing API Server certificate and key.")
+       } else {
+               // The certificate and the key did NOT exist, let's generate them now
+               // TODO: Add a test case to verify that this cert has the x509.ExtKeyUsageClientAuth flag
+               config := certutil.Config{
+                       CommonName: "kube-apiserver-kubelet-client",
+                       Organization: []string{"system:masters"},
+                       Usages:       []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+               }
+               apiClientCert, apiClientKey, err := pkiutil.NewCertAndKey(caCert, caKey, config)
+               if err != nil {
+                       return fmt.Errorf("failure while creating API server client key and certificate [%v]", err)
+               }
+
+               if err = pkiutil.WriteCertAndKey(pkiDir, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName, apiClientCert, apiClientKey); err != nil {
+                       return fmt.Errorf("failure while saving API server client certificate and key [%v]", err)
+               }
+               fmt.Println("[certificates] Generated API server certificate and key.")
+       }
+
        fmt.Printf("[certificates] Valid certificates and keys now exist in %q\n", pkiDir)
 
        return nil
```